flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java

```diff
@@ -0,0 +1,812 @@
+        RelNode newInput = match.getInput().accept(this);
+        RexTimeIndicatorMaterializer materializer = new RexTimeIndicatorMaterializer(newInput);
+
+        Function<Map<String, RexNode>, Map<String, RexNode>> materializeExprs =
+                rexNodesMap ->
+                        rexNodesMap.entrySet().stream()
+                                .collect(
+                                        Collectors.toMap(
+                                                Map.Entry::getKey,
+                                                e -> e.getValue().accept(materializer),
+                                                (e1, e2) -> e1,
+                                                LinkedHashMap::new));
+        // update input expressions
+        Map<String, RexNode> newPatternDefs = materializeExprs.apply(match.getPatternDefinitions());
+        Map<String, RexNode> newMeasures = materializeExprs.apply(match.getMeasures());
+        RexNode newInterval = null;
+        if (match.getInterval() != null) {
+            newInterval = match.getInterval().accept(materializer);
+        }
+
+        Predicate<String> isNoLongerTimeIndicator =
+                fieldName -> {
+                    RexNode newMeasure = newMeasures.get(fieldName);
+                    if (newMeasure == null) {
+                        return false;
+                    } else {
+                        return !isTimeIndicatorType(newMeasure.getType());
+                    }
+                };
+
+        // decide the MATCH_ROWTIME() return type is TIMESTAMP or TIMESTAMP_LTZ, if it is
+        // TIMESTAMP_LTZ, we need to materialize the output type of LogicalMatch node to
+        // TIMESTAMP_LTZ too.
+        boolean isLtzRowtimeIndicator =
+                newMeasures.values().stream()
+                        .anyMatch(
+                                node ->
+                                        isTimestampLtzIndicatorType(node.getType())
+                                                && isRowtimeIndicatorType(node.getType()));
+        // materialize all output types
+        RelDataType newOutputType =
+                getRowTypeWithoutTimeIndicator(
+                        match.getRowType(), isLtzRowtimeIndicator, isNoLongerTimeIndicator);
+        return new LogicalMatch(
+                match.getCluster(),
+                match.getTraitSet(),
+                newInput,
+                newOutputType,
+                match.getPattern(),
+                match.isStrictStart(),
+                match.isStrictEnd(),
+                newPatternDefs,
+                newMeasures,
+                match.getAfter(),
+                match.getSubsets(),
+                match.isAllRows(),
+                match.getPartitionKeys(),
+                match.getOrderKeys(),
+                newInterval);
+    }
+
+    @Override
+    public RelNode visit(RelNode node) {
+        if (node instanceof Collect) {
```

Comment:
one suggestion is: we can extend RelShuttle and add `RelNode visit(Collect collect)`, `RelNode visit(LogicalWindowAggregate agg), etc..` to new shuffle, this class can extend from the new shuffle and avoid so many `if else`

Reply:
If we introduce a new interface whichs extend `RelShuttle` like the interface `HiveRelShuttle`, We need override `accept` method in `RelNode` just like the `HiveFilter`, `HiveJoin`.
Do we really need to do this?
BTW, If we introduce a new interface which now matches `LogicalNode`, after we refactor `RelTimeIndicatorConverter` to work just before `physical_transform` phase, we need update the new  `RelShuttle` interface to match `FlinkLogicalNode` and update those `FlinkLogicalNode` to override `accept` method.

Reply:
make sense